app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/chart/BeaconLineChart.java

```diff
@@ -317,74 +321,75 @@ protected void drawBeacons(Canvas canvas) {
-            }
+            drawCircleForPreviousPoint(canvas);
+            drawNextPoint(canvas, advertisingPacket);
+        }
+        currentBeaconIndex++;
+        fadeLastPoint(canvas);
+    }
+
+    protected void drawCircleForPreviousPoint(Canvas canvas) {
+        canvas.drawCircle(currentLinePoint.x, currentLinePoint.y, pixelsPerDip * 1.5f, linePaint);
+    }
+
+    protected void prepareDraw(Beacon beacon) {
+        lastLinePoint = null;
+        lastAdvertisingPacket = null;
+
+        linePaint.setShader(createLineShader(getBeaconColor(beacon, coloringMode, currentBeaconIndex)));
+        linePaint.setAlpha(128);
+    }
 
-            lastLinePoint.x = currentLinePoint.x;
-            lastLinePoint.y = currentLinePoint.y;
-            lastAdvertisingPacket = advertisingPacket;
-            currentBeaconIndex++;
+    protected void drawNextPoint(Canvas canvas, AdvertisingPacket advertisingPacket) {
+        if (lastLinePoint != null && lastAdvertisingPacket != null) {
+            if (advertisingPacket.getTimestamp() < lastAdvertisingPacket.getTimestamp() + TimeUnit.SECONDS.toMillis(5)) {
+                canvas.drawLine(
+                        lastLinePoint.x,
+                        lastLinePoint.y,
+                        currentLinePoint.x,
+                        currentLinePoint.y,
+                        linePaint
+                );
+            }
+        } else {
+            lastLinePoint = new PointF();
         }
 
+        lastLinePoint.x = currentLinePoint.x;
+        lastLinePoint.y = currentLinePoint.y;
+        lastAdvertisingPacket = advertisingPacket;
+    }
+
+    protected void fadeLastPoint(Canvas canvas) {
         if (lastLinePoint != null) {
             linePaint.setAlpha(255);
             canvas.drawCircle(lastLinePoint.x, lastLinePoint.y, pixelsPerDip * 4, linePaint);
         }
     }
 
-    protected float getValue(Beacon beacon, AdvertisingPacket advertisingPacket, long windowLength) {
-        List<AdvertisingPacket> recentAdvertisingPackets = null;
-        int[] recentRssis;
-        float meanRssi;
-
-        // make sure that the window size is at least 10 seconds when we're looking for the frequency
-        windowLength = (valueType != VALUE_TYPE_FREQUENCY) ? windowLength : Math.max(windowLength, 10000);
-
-        if (windowLength == 0) {
-            meanRssi = advertisingPacket.getRssi();
-        } else {
-            recentAdvertisingPackets = beacon.getAdvertisingPacketsBetween(
-                    advertisingPacket.getTimestamp() - windowLength,
-                    advertisingPacket.getTimestamp()
-            );
-            recentRssis = AdvertisingPacketUtil.getRssisFromAdvertisingPackets(recentAdvertisingPackets);
-            meanRssi = AdvertisingPacketUtil.getMeanRssi(recentRssis);
-        }
+    protected float getValue(Beacon beacon, AdvertisingPacket advertisingPacket) {
+        int rssi = advertisingPacket.getRssi();
 
         switch (valueType) {
```

Comment:
Switch case is equivalent in the this class and `RssiFilterLineChart`. Reduce Code duplication.